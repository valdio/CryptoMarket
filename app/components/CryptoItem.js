import React from 'react'
import {View, Text, StyleSheet, Image} from 'react-native'
import {COLORS, Spacing, Typography} from '../template'
import rate_down from '../assets/images/rate/rate_down.png'
import rate_up from '../assets/images/rate/rate_up.png'
import LinearGradient from 'react-native-linear-gradient'
import {Link} from 'react-router-native'
import {ROUTE} from '../lib/routing/routes'

export default CryptoItem = ({item} = props) => {
  const quoteUSD = item && item.quote && item.quote.USD || {}//inline validation
  return (<Link to={`${ROUTE.CRYPTO_DETAILS_TAG}/${item.id}`}
                underlayColor={COLORS.BACKGROUND}><View style={styles.container}>
    <View style={styles.content}>
      <View style={styles.symbolContainer}>
        <Text style={styles.symbolText}>{item.symbol}</Text>
      </View>
      <Text style={styles.cryptoName}>{item.name}</Text>
      <View style={styles.statsHolder}>
        {renderQuoteChange(quoteUSD.percent_change_1h, 'hour')}
        {renderQuoteChange(quoteUSD.percent_change_24h, 'day')}
        {renderQuoteChange(quoteUSD.percent_change_7d, 'week')}
      </View>
    </View>
    <LinearGradient colors={[COLORS.TRANSPARENT, COLORS.LIGHT_BLUE, COLORS.BLUE]} style={styles.linearGradient}/>
  </View>
  </Link>)
}

const renderQuoteChange = (percentage, percentageDetails) => !percentage ? <View/> :
  <View style={styles.quoteContainer}>
    <Image style={styles.quoteImage} source={percentage < 0 ? rate_down : rate_up}/>
    <View style={{alignItems: 'center'}}>
      <Text stule={styles.quotePercentage}>{percentage.toFixed(2)}%</Text>
      <Text style={styles.quotePercentageDetails}>$/{percentageDetails}</Text>
    </View>
  </View>


const styles = StyleSheet.create({
  container: {
    paddingVertical: 5,
    marginBottom: 12,
    marginHorizontal: 10
  },
  content: {
    marginHorizontal: 10,
    flexDirection: 'row',
    alignItems: 'center'
  },
  symbolContainer: {
    backgroundColor: COLORS.LIGHT_BLUE,
    width: 50,
    height: 50,
    borderRadius: 25,
    justifyContent: 'center',
    alignItems: 'center'
  },
  symbolText: {
    ...Typography.boldText,
    color: COLORS.BLUE_TEXT
  },
  cryptoName: {
    ...Typography.sectionTitle,
    marginHorizontal: 6

  },
  statsHolder: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    flexDirection: 'row'
  },

  //quote change
  quoteContainer: {
    marginHorizontal: 5,
    alignItems: 'center',
    justifyContent: 'center'
  },
  quotePercentage: {
    fontSize: 12,
    color: COLORS.LIGHT_BLUE
  },
  quoteImage: {
    width: 20,
    height: 20,
    resizeMode: 'contain'
  },
  quotePercentageDetails: {
    color: COLORS.LIGHT_BLUE,
    fontSize: 10
  },

  linearGradient: {
    position: 'absolute',
    top: 0, bottom: 0,
    flex: 1,
    right: 0, left: 0,
    borderRadius: 6,
    opacity: .1
  }
})

