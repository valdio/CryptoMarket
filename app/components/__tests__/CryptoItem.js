import React from 'react'
import renderer from 'react-test-renderer'
import CryptoItem from '../CryptoItem'
import {cryptoCoinMock, cryptoMetadataMock} from './mock'
import {shallow} from 'enzyme'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {Link} from 'react-router-native'
import LinearGradient from 'react-native-linear-gradient'

Enzyme.configure({adapter: new Adapter()})
const wrapper = shallow(<CryptoItem item={cryptoCoinMock}/>)

it('renders CryptoItem using Snapshots', () => {
  const tree = renderer.create(<CryptoItem item={cryptoCoinMock}/>)
  expect(tree).toMatchSnapshot()
})

test('CryptoItem is a react-router-native Link', () => {
  expect(wrapper.find(Link).length).toBe(1)
})

describe('CryptoItem has title', () => {
  expect(wrapper.findWhere(node => node.prop('testID') === 'card-title').text()).toBeTruthy()
})

describe('CryptoItem has symbol', () => {
  expect(wrapper.findWhere(node => node.prop('testID') === 'card-symbol').text()).toBeTruthy()
})

describe('CryptoItem has LinearGradient', () => {
  expect(wrapper.find(LinearGradient).length).toBe(1)
})

