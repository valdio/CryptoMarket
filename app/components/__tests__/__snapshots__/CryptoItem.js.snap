// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CryptoItem using Snapshots 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "marginBottom": 12,
        "marginHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginHorizontal": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#AAC4F8",
            "borderRadius": 25,
            "height": 50,
            "justifyContent": "center",
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#404D66",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
          testID="card-symbol"
        >
          BTC
        </Text>
      </View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#484848",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 6,
          }
        }
        testID="card-title"
      >
        Bitcoin
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 5,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/rate/rate_down.png",
              }
            }
            style={
              Object {
                "height": 20,
                "resizeMode": "contain",
                "width": 20,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              stule={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 12,
                }
              }
            >
              -0.15
              %
            </Text>
            <Text
              style={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 10,
                }
              }
            >
              $/
              hour
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 5,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/rate/rate_up.png",
              }
            }
            style={
              Object {
                "height": 20,
                "resizeMode": "contain",
                "width": 20,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              stule={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 12,
                }
              }
            >
              0.52
              %
            </Text>
            <Text
              style={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 10,
                }
              }
            >
              $/
              day
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginHorizontal": 5,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/rate/rate_up.png",
              }
            }
            style={
              Object {
                "height": 20,
                "resizeMode": "contain",
                "width": 20,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              stule={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 12,
                }
              }
            >
              0.99
              %
            </Text>
            <Text
              style={
                Object {
                  "color": "#AAC4F8",
                  "fontSize": 10,
                }
              }
            >
              $/
              week
            </Text>
          </View>
        </View>
      </View>
    </View>
    <BVLinearGradient
      colors={
        Array [
          0,
          4289381624,
          4282223615,
        ]
      }
      endPoint={
        Object {
          "x": 0.5,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0.5,
          "y": 0,
        }
      }
      style={
        Object {
          "borderRadius": 6,
          "bottom": 0,
          "flex": 1,
          "left": 0,
          "opacity": 0.1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
</View>
`;
